src/wasm-validator.h

```diff
@@ -46,188 +48,8 @@
-  void visitSwitch(Switch *curr);
-  void visitCall(Call *curr);
-  void visitCallImport(CallImport *curr);
-  void visitCallIndirect(CallIndirect *curr);
-  void visitGetLocal(GetLocal* curr);
-  void visitSetLocal(SetLocal *curr);
-  void visitLoad(Load *curr);
-  void visitStore(Store *curr);
-  void visitAtomicRMW(AtomicRMW *curr);
-  void visitAtomicCmpxchg(AtomicCmpxchg *curr);
-  void visitAtomicWait(AtomicWait *curr);
-  void visitAtomicWake(AtomicWake *curr);
-  void visitBinary(Binary *curr);
-  void visitUnary(Unary *curr);
-  void visitSelect(Select* curr);
-  void visitDrop(Drop* curr);
-  void visitReturn(Return* curr);
-  void visitHost(Host* curr);
-  void visitImport(Import* curr);
-  void visitExport(Export* curr);
-  void visitGlobal(Global* curr);
-  void visitFunction(Function *curr);
-
-  void visitMemory(Memory *curr);
-  void visitTable(Table* curr);
-  void visitModule(Module *curr);
-
-  void doWalkFunction(Function* func) {
-    PostWalker<WasmValidator>::doWalkFunction(func);
-  }
-
-  // helpers
- private:
-  template <typename T, typename S>
-  std::ostream& fail(S text, T curr);
-  std::ostream& printFailureHeader();
-
-  template<typename T>
-  bool shouldBeTrue(bool result, T curr, const char* text) {
-    if (!result) {
-      fail("unexpected false: " + std::string(text), curr);
-      return false;
-    }
-    return result;
-  }
-  template<typename T>
-  bool shouldBeFalse(bool result, T curr, const char* text) {
-    if (result) {
-      fail("unexpected true: " + std::string(text), curr);
-      return false;
-    }
-    return result;
-  }
-
-  template<typename T, typename S>
-  bool shouldBeEqual(S left, S right, T curr, const char* text) {
-    if (left != right) {
-      std::ostringstream ss;
-      ss << left << " != " << right << ": " << text;
-      fail(ss.str(), curr);
-      return false;
-    }
-    return true;
-  }
-
-  template<typename T, typename S>
-  bool shouldBeEqualOrFirstIsUnreachable(S left, S right, T curr, const char* text) {
-    if (left != unreachable && left != right) {
-      std::ostringstream ss;
-      ss << left << " != " << right << ": " << text;
-      fail(ss.str(), curr);
-      return false;
-    }
-    return true;
-  }
-
-  template<typename T, typename S>
-  bool shouldBeUnequal(S left, S right, T curr, const char* text) {
-    if (left == right) {
-      std::ostringstream ss;
-      ss << left << " == " << right << ": " << text;
-      fail(ss.str(), curr);
-      return false;
-    }
-    return true;
-  }
-
-  void shouldBeIntOrUnreachable(WasmType ty, Expression* curr, const char* text);
-  void validateAlignment(size_t align, WasmType type, Index bytes, bool isAtomic,
-                         Expression* curr);
-  void validateMemBytes(uint8_t bytes, WasmType type, Expression* curr);
-  void validateBinaryenIR(Module& wasm);
+struct WasmValidator {
+  bool validate(Module& module, bool validateWeb = false, bool validateGlobally = true, bool quiet = false);
```

Comment:
Three bool args make me think this should be flags.
```c++
enum ValidatorFlags {
  ValidateWeb = 1 << 0,
  ValidateGlobally = 1 << 1,
  ValidateQuiet = 1 << 2
}
//...
bool WasmValidator::validate(Module& module, ValidatorFlags flags);
//...
validator.flags(wasm, ValidateGlobally | ValidateQuiet);
```

Reply:
Yeah, makes sense. How about if I do that in a followup? (I'm trying to change the API as little as possible in this PR)

Reply:
Sounds reasonable